internal/operators/operators.go

```diff
@@ -0,0 +1,301 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operators
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/inspektor-gadget/inspektor-gadget/internal/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/internal/parser"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+type Runner interface {
+	ID() string
+	Parser() parser.Parser
+	Gadget() gadgets.Gadget
+	Context() context.Context
+	Operators() Operators
+	Logger() logger.Logger
+}
+
+type (
+	EnricherFunc func(any) error
+)
+
+var ErrAsync = errors.New("async")
+
+type Operator interface {
+	// Name must return a unique name for the operator
+	Name() string
+
+	// Description is an optional description to show to the user
+	Description() string
+
+	// Params will return global params (required) for this operator
+	Params() params.ParamDescs
+
+	// PerGadgetParams will return params (required) per gadget instance of the operator
+	PerGadgetParams() params.ParamDescs
+
+	// Dependencies can list other allOperators that this operator depends on
+	Dependencies() []string
+
+	// CanOperateOn should test whether the operator supports the given gadget. Init has not
+	// necessarily been called at this point.
+	CanOperateOn(gadgets.Gadget) bool
+
+	// Init allows the operator to initialize itself
+	Init(params *params.Params) error
+
+	// Close allows the operator to clean up stuff prior to exiting
+	Close() error
+
+	// Instantiate is called before a gadget is run (before PreGadgetRun) with this operator
+	// This must return something that implements operator as well.
+	// This is useful to create a context for an operator by wrapping it.
+	// Params given here are the ones returned by PerGadgetParams()
+	Instantiate(runner Runner, gadgetInstance any, perGadgetParams *params.Params) (OperatorInstance, error)
+}
+
+type OperatorInstance interface {
+	// PreGadgetRun is called before a gadget is started but after all allOperators have been initialized
+	PreGadgetRun() error
+
+	// PostGadgetRun is called on the operator that was returned from PrepareTrace after a
+	// gadget was stopped
+	PostGadgetRun() error
+
+	// Enricher should return a function that is able to operator on the event ev and
+	// call next afterwards.
+	Enricher(next EnricherFunc) EnricherFunc
+
+	// Deprecated: EnrichEvent
+	EnrichEvent(ev any) error
+}
+
+type operatorInterfaces struct {
+	Operator
+	OperatorInstance
+}
+
+type Operators []operatorInterfaces
```

Comment:
I'm not a big fan of this. I think the code shouldn't allow the developer to call `operator.{PreGadgetRun|PostGadgetRun|Enrich|Enricher}` without calling `Instantiate`.

I think I found a middle-ground for this: https://github.com/inspektor-gadget/inspektor-gadget/commit/074ccd4fd1ef626163e0d936701ea7fcd288ae25. I keeps `Operator` and `OperatorInstance` but doesn't mix them. `OperatorInstance` would be only accessible if `Operator.Instantiate` is called. WDYT?

cc @mauriciovasquezbernal

Reply:
That's much better!

Reply:
IMO it goes on the right direction. One difference is that we're not saving the operator instances explicitly, but I don't think it's a problem.

Reply:
> One difference is that we're not saving the operator instances explicitly, but I don't think it's a problem.
>
Doing so, it is even more evident that the lifecycle of an `OperatorInstance` is tied to the gadget lifecycle (`runGadget` function), don't you think?

Reply:
Yes, that's true. I was thinking that there could be cases where the `Operator` wants to keep track of all `OperatorInstances` but this can be implemented on each operator if needed.